models/exchange/binance/api.py

```diff
@@ -67,186 +72,327 @@ def __init__(self, api_key: str='', api_secret: str='', api_url: str='https://ap
 
         if len(df) == 0:
             return pd.DataFrame()
 
-        # fix: float division by zero
-        # remove pending orders with status new (i.e. order made by user at limit)
-        df = df[df['status'] == "FILLED"]
-
-        # replace null NaN values with 0
-        df.fillna(0, inplace=True)
-
-        df = df[[ 'time', 'symbol', 'side', 'type', 'executedQty', 'cummulativeQuoteQty', 'status' ]]
-        df.columns = [ 'created_at', 'market', 'action', 'type', 'filled', 'size', 'status' ]
-        df['created_at'] = df['created_at'].apply(lambda x: int(str(x)[:10]))
-        df['created_at'] = df['created_at'].astype("datetime64[s]")
-        df['size'] = df['size'].astype(float)
-        df['filled'] = df['filled'].astype(float)
-        df['action'] = df['action'].str.lower()
-        df['type'] = df['type'].str.lower()
+        # feature engineering
+
+        def convert_time(epoch: int=0):
+            epoch_str = str(epoch)[0:10]
+            return datetime.fromtimestamp(int(epoch_str))
+
+        df.time = df['time'].map(convert_time)
+        df['time'] = pd.to_datetime(df['time']).dt.tz_localize('UTC')
+
+        df['size'] = np.where(df['side']=='BUY', df['cummulativeQuoteQty'], np.where(df['side']=='SELL', df['executedQty'], 222))
+        df['fees'] = df['size'].astype(float) * 0.001
+        df['fees'] = df['fees'].astype(object)
+
+        df['side'] = df['side'].str.lower() 
+
+        df.rename(columns={ 
+            'time': 'created_at', 
+            'symbol': 'market', 
+            'side': 'action', 
+            'executedQty': 'filled' 
+        }, errors='raise', inplace=True)
+
+        def convert_status(status: str=''):
+            if status == 'FILLED':
+                return 'done'
+            elif status == 'NEW':
+                return 'open'
+            elif status == 'PARTIALLY_FILLED':
+                return 'pending'
+            else:
+                return status    
+
+        df.status = df.status.map(convert_status)
         df['status'] = df['status'].str.lower()
-        df['price'] = df['size'] / df['filled']
 
-        # pylint: disable=unused-variable
-        for k, v in df.items():
-            if k == 'status':
-                df[k] = df[k].map(self.__convertStatus)
+        # select columns
+        df = df[[ 'created_at', 'market', 'action', 'type', 'size', 'filled', 'fees', 'price', 'status' ]]
 
+        # filtering
         if action != '':
             df = df[df['action'] == action]
-            df = df.reset_index(drop=True)
-
-        if status != 'all' and status != '':
+        if status != 'all':
             df = df[df['status'] == status]
-            df = df.reset_index(drop=True)
 
         return df
 
+    def getTime(self) -> datetime:
+        """Retrieves the exchange time"""
+    
+        def convert_time(epoch: int=0):
+            epoch_str = str(epoch)[0:10]
+            return datetime.fromtimestamp(int(epoch_str))
```

Comment:
this method was defined already in lines 292-294, we can just use that one, no?

Reply:
Ok, will move it out. Thanks

Reply:
This is done. Thanks